### ISSUE TITLE
Currupt backup cannot be repaired

### ISSUE THREAD

**Issue by BlackEdder0815** [duplicati/duplicati#3037]
- [/] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.2.19_canary_2018-02-12 
- **Operating system**: Debian ARM
- **Backend**: WebDAV

## Description

I'm using Duplicati now for over a year with changi
changing backends. With the last change, I decided to recreate the whole database.
After recreating, I got some errors about some missing remote files.
A repair say: repair not possible, I should try list and purge-broken-files.
But list-broken-files says simply, there are no broken files.
But verify brings the message:

```Listing remote folder ...
Missing file: duplicati-b4c3d40ad78e4413ab710fd7e16324922.dblock.zip.aes
Missing file: duplicati-b35ae43ae81054328a64637cf83e56a68.dblock.zip.aes
Missing file: duplicati-b88f6126905784408ac0ae596c362aac6.dblock.zip.aes
Missing file: duplicati-bade5355cd40b43bb8364ff38a7e283a8.dblock.zip.aes
Found 4 files that are missing from the remote storage, please run repair

Found 4 files that are missing from the remote storage, please run repair
Return code: 100
```

Do you have any suggestions,  how to go on here?

**Comment by wjansenw** [comment]
Are those files available in the backend? What backend are you using? For some backend ( Google drive I believe), duplicati can only see the files it crrated.

**Comment by wjansenw** [comment]
I see the backend is webdav. Please verify if the files exist.  Maybe they were not copied when changing backend.

**Comment by BlackEdder0815** [comment]
No, the files are not existing in the backend. Wether in the new, nor in the old. But this is not the point. Duplicati seems to be in a cycle and can't break out. It's currently not  possible to repair the backup (restore the missing files) within duplicati.
Or I'm not aware about the correct steps...

**Comment by wjansenw** [comment]
In that case it will probably be a dindex file that had not been created or
deleted correctly. Try to locate the dindex files that refer to the dblock
files. You will have to dive into the sqlite database. Move those dindex
files outside your backup location and do a repair (or recreate even).

This should solve it I believe. But that doesn't give an answer to what
brought your backup in this state.

**Comment by BlackEdder0815** [comment]
@wjansenw good job!
I searched in the table "Remotevolume" for the related file, went with this ID to the table "IndexBlockLink" and filtered for the related BlockVolumeID, took the listed IndexVolumeID and searched in the table "Remotevolume" for this ID. The related file should be a dindex-file. I moved this out of my backup and recreated the database.
Then, I got no error about any issues.

Why was this happened? I really don't know. I used the ACD (Amazon Cloud Drive) backend right before, but since beginning of 2017. In the beginning, I had some trouble with uploading, with crashing duplicati and other issues. Maybe this was an relict of this time? The time stamp of the dindex-files are all from beginning of 2017. So maybe this is no issue of the actual version of duplicati!

**Comment by BlackEdder0815** [comment]
Maybe you should think about to solve such issues automatically?

**Comment by BlackEdder0815** [comment]
For full documentation (if anyone have also this issue) use the following query to get all files, which have to be deleted:
```
select R1.Name from Remotevolume as R join IndexBlockLink as I on R.ID = I.BlockVolumeID join Remotevolume as R1 on I.IndexVolumeID = R1.ID where R.Name in (
"duplicati-bf4c5c1d47a614e1f9ac8aa02f609e68f.dblock.zip.aes",
"duplicati-b161b95769f5c47298c58048ebf04ad46.dblock.zip.aes",
"...")
```

**Comment by kenkendk** [comment]
Awesome debugging!
At least knowning what causes the problem (and how to fix it) is valuable.

We can add a handler to the re-create/repair steps that looks at this, and auto-fixes it.

But hopefully, this will also make it possible to guess why it fails.

@BlackEdder0815 If you did a recreate the remote operations are gone, but just in case you did a repair: can you check the `RemoteOperation` table for mentions of the `dindex` file that you had to remove?

**Comment by BlackEdder0815** [comment]
I did a recreate, but I have a copy of the database.
I will have a look and share the results tomorrow

**Comment by BlackEdder0815** [comment]
Hm, nothing found, what was helpful.
Only a mention get-operation:
```
"242"	"1"	"1518710962"	"get"	"duplicati-i13dd35c6dc92430a94f1d199240b73f5.dindex.zip.aes"	"{""Size"":36237,""Hash"":""s/+1/KkmRbVewUvPg1uUCxZytYE6x/Ji2+viPgpp4ig=""}"
```
Of course, I had maybe also in the past made a recreate of the database, don't know. The related dblock-file was not mentioned in the table "RemoteOperation". So I really don't know how the backup get into this state...

**Comment by tygill** [comment]
I've also been having some issues with attempting to repair / recreate a database, though my issues are mainly that the repair process is taking an extremely long time to run (when running it overnight on my laptop it kept failing with 'TaskCancelledException' possibly when the laptop went to sleep or overheated).

Yesterday I started a repair on a desktop (using 2.0.3.6), which has now been running for nearly 24 hours. Looking at the console log (and the sizes of the files it has downloaded in), it appears to have first downloaded the 23 dlist files, then the 1040 dindex files, and is currently attempting to download all 250GB of dblock files (it has currently completed ~449 of 1040 files). Based on the [documentation here](https://github.com/duplicati/duplicati/wiki/Developer-Documentation#without-the-local-database), my understanding is that for repair, usually only the dlist and dindex files should be required. As I believe the remote backup data is in a good, clean state (particularly since there are 1040 of both the dindex and dblock files, and I have no reason to believe that anything has been deleted or changed there), this is particularly surprising.

I tried reading through the code (particularly, RecreateDatabaseHandler.DoRun, lines 398-467, where it seems to be picking the dblock volumes that need to be downloaded and processed) to try to understanding under what circumstances the dindex / dlist files aren't sufficient for recreate / repair, but don't understand enough of the database schema to have picked it out yet.

**Comment by SamiLehtinen** [comment]
Today I made some additional tests and I find the results very alarming. 

**Covertly broken backups, restore fails, otherwise everything seems to be ok**

Running: repair, test, purge-broken-files, backup -> ok
Running: restore -> says it's successful, but in reality no files were restored. Yet there's a mention about missing files, after the local database has been reconstructed using remote files.

Steps to reproduce?

 - Generate backups as usual
 - Delete one backup set, let's say the "last one". 
I haven't tried different corruption models yet. I can do that intentionally corrupting files in different ways after the basic stuff works. Current corruption model is the simplest one. Just one incremental backup set went completely missing.
 - Run purge-broken-files and repair -> After this everything **should** be fine again.
 - Run more backups, works as usual.
 - Run restore, for the latest backup run -> **Fail**

This kind of functionality is very dangerous. Now I got unknown number of backups, which seem to be officially in good state, but in reality those are totally useless. If I would be running operations as most do, I would be just happy. All is well.

I haven't went into details, is the problem actually in the repair or reconstruction code. I think it's the repair, shouldn't it remove all invalid references from the data set (?). 

It's also possible, that the restore from repository function which rebuilds the database for restore, does include some invalid references in the process and therefore the restore isn't working.

If required, I can provide mock-up data set to repeat the restore fail easily. Yet based on the steps to reproduce it shouldn't be too hard.

I've got also different automated testing. So when basic stuff is working. I can leave systems to do some fuzzing corruption for a weekend to see what kind of mess we then get to recover.

**Comment by instantlinux** [comment]
Ditto all the above. I have been recommending duplicati to other users but given my own experiences with failed metadata database events over the past year, I'm now contemplating a move away from Duplicati. Is there anyone working on this? I'd work on it myself but Duplicati is written in a language that I have no experience with.

**Comment by drwtsn32x** [comment]
> failed metadata database events

Can you elaborate on that?

If you are willing, please try the latest experimental version (2.0.5.0). In my experience it has solved all the random issues I had with the previous versions - unexpected difference in fileset, long database recreation, etc.

There is still a known issue when the "stop now" function is used (and presumably if you kill Duplicati) where it can cause a "detected non-empty blocksets" error, but that is being actively worked on.
